docs/design: add a proposal for pausing user DDL when upgrading DDL (#43061)
pingcap/tidb: docs/design/2023-04-11-pause-user-ddl-when-upgrading.md

* Author: [zimulala](https://github.com/zimulala)
* Tracking issue: https://github.com/pingcap/tidb/issues/39751

## Abstract

This document describes a feature that allows users to pause the execution of their Data Definition Language (DDL) statements during the upgrade process of TiDB. During the upgrade process, the system stops the execution of user DDL statements, completes the system DDL operation, and finally resumes the execution of the original user DDL statements. This feature is available for upgrades from v7.1 and later versions of TiDB, and it does not support upgrades from versions prior to v7.1.

## Motivation or Background

Since the cluster upgrade may need to deal with DDL statements(there may be some DDL statements being processed before the upgrade), and DDL itself implementation framework will also need to adjust, resulting in some versions may not be able to roll the upgrade situation. Although these cases only exist in some versions, it is not easy to describe them one by one. Therefore, when upgrading TiDB clusters, users need to confirm that there are no DDL statements being executed, otherwise there may be problems with undefined behavior.

> Now there are two questions:

Reviewer: ```suggestion
Now there are two problems:
```

> - The above rule is only described in the doc, and users may misoperate. Related instructions doc.

Reviewer: What does "Related instructions doc" mean?

- During the upgrade operation, users need to communicate and coordinate with each business to cancel DDL, then upgrade, and finally re-execute DDL. Specific scenarios such as large cluster scenarios shared across multiple systems, such as scenarios on TiDB cloud.

## Detailed Design

Here are some roles for TiDB in the cluster (assuming upgrading from v7.1 to v7.2 here):
- New TiDB: The new version of TiDB that will be upgraded, such as v7.2.
- Old TiDB: Other TiDBs in the cluster that have not been upgraded, such as v7.1.
- Owner TiDB: The TiDB where the DDL owner in the cluster is located.

Specific upgrade process:
1. New TiDB: Check whether the old TiDB version is greater than or equal to v7.1, if not, then exit the whole process.
2. Notify the cluster that all TiDB enters to pause executing user DDL operations. Notification method:
   1. New TiDB: Set the PD path (for example: `/tidb/server/global_state` , use this path later) to the current TiDB version. In addition, in order to prevent multiple TiDB nodes from upgrading at the same time, it will check whether the value on the `/tidb/server/global_state` path is empty.
   2. All TiDBs：

>       1. The newly started TiDB will first check the `/tidb/server/global_state` value and whether it needs to enter upgrading_mode mode.

Reviewer: ```suggestion
      1. The newly started TiDB will first check the `/tidb/server/global_state` value and whether it needs to enter the upgrading mode.
```

>       2. The running TiDB will watch `/tidb/server/global_state` , and when it receives an upgrade status notification, it will enter upgrading_mode mode.

Reviewer: ```suggestion
      2. The running TiDB will watch `/tidb/server/global_state` , and when it receives an upgrade status notification, it will enter the upgrading mode.
```

>    3. Owner TiDB: Set upgrading in the owner value on the PD.

Reviewer: Could you describe this in detail? 

What does the "upgrading" mean? Is it a flag? What does the "owner value" mean? Is it related to any etcd key?

  - During the upgrade process, the operation of some components may be blocked, such as:
    - DM, TiCDC and Lightning (Logical Import Mode) and other components. That is, during the upgrade process, import SQL to TiDB through such components, and there are DDL operations (if there is a system DDL operation that may have latent risk), then these DDL will be set to paused, causing this import operation to block.

## Test Design

- The original upgrade tests.
- Test some of the following scenarios to check whether the upgrade is normal, and whether the DDL operation is as expected.
  - Add DDL operations before upgrading.
  - Test scenarios during the upgrade process:
    - Add DDL operations.
    - The upgrade was interrupted.
    - The DDL owner has changed.

## Impacts & Risks

Since TiDB itself has some functional roles or judgments on whether some functions are enabled or not, and there will be different versions and other factors during the upgrade process, we need to consider more scenarios, and these scenarios may have omissions in design or testing.

> Investigation & Alternatives

Reviewer: ```suggestion

## Investigation & Alternatives

```

# Proposal: Pause User DDL when Upgrading TiDB

* Author: [zimulala](https://github.com/zimulala)
* Tracking issue: https://github.com/pingcap/tidb/issues/39751

## Abstract

This document describes a feature that allows users to pause the execution of their Data Definition Language (DDL) statements during the upgrade process of TiDB. During the upgrade process, the system stops the execution of user DDL statements, completes the system DDL operation, and finally resumes the execution of the original user DDL statements. This feature is available for upgrades from v7.1 and later versions of TiDB, and it does not support upgrades from versions prior to v7.1.

## Motivation or Background

> Since the cluster upgrade may need to deal with DDL statements(there may be some DDL statements being processed before the upgrade), and DDL itself implementation framework will also need to adjust, resulting in some versions may not be able to roll the upgrade situation. Although these cases only exist in some versions, it is not easy to describe them one by one. Therefore, when upgrading TiDB clusters, users need to confirm that there are no DDL statements being executed, otherwise there may be problems with undefined behavior.

Reviewer 2: ```suggestion
1. Since the TiDB cluster upgrade may need to deal with DDL statements(there may be some DDL statements being processed before the upgrade).
2. DDL module itself implementation logic will also need to be adjusted, resulting in some versions may not allow user DDL job running during roll upgrade period. 
Although above scenarios may only exist in some upgrade cases, it is not easy to describe them one by one. Therefore, when upgrading TiDB clusters, users need to first confirm that there are no DDL statements being executed, otherwise there may be problems with unexpection behavior.
```

Now there are two problems:
- The above rule is only described in the doc, and users may misoperate. Related [doc](https://docs.pingcap.com/tidb/dev/upgrade-tidb-using-tiup#upgrade-tidb-using-tiup).
- During the upgrade operation, users need to communicate and coordinate with each business to cancel DDL, then upgrade, and finally re-execute DDL. Specific scenarios such as large cluster scenarios shared across multiple systems, such as scenarios on TiDB cloud.

## Detailed Design

Here are some roles for TiDB in the cluster (assuming upgrading from v7.1 to v7.2 here):
- New TiDB: The new version of TiDB that will be upgraded, such as v7.2.
- Old TiDB: Other TiDBs in the cluster that have not been upgraded, such as v7.1.
- Owner TiDB: The TiDB where the DDL owner in the cluster is located.

Specific upgrade process:

> 1. New TiDB: Check whether the old TiDB version is greater than or equal to v7.1, if not, then exit the whole process.

Reviewer 2: ```suggestion
1. New TiDB: Check whether the old TiDB version is greater than or equal to v7.1, if not, then exit the whole process. That means the cluster upgrade process will follow legcy upgrade ways, that means User have to make sure there is no user issued DDL in running state.
```

4. New TiDB: Perform upgrade-related operations in normal bootstrap.
5. New TiDB: Restore the original user DDL operation, that is, resume all DDL operations (the operation that the user paused before bootstrap needs to be preserved).
6. Resume normal execution of DDL mode
   1. New TiDB: Release pause mode on PD and set `/tidb/server/global_state` and the owner value to normal state.
   2. All TiDBs: watch `/tidb/server/global_state` , receive this notification, and enter normal mode.
7. Complete the bootstrap operation.

![Figure 1: step 2 flow chart](./imgs/pause-user-ddl-when-upgrading-step-2.png)

### Limitations

In the plan, we distinguish between DDL operation types. Among them, we distinguish whether the DDL operation is a system DDL operation or a user DDL operation  by whether the DDL operation is performed on the system table. Therefore, we require users to not perform DDL operations on system tables when upgrading.

## Compatibility

- TiDB
  - This feature is not supported when upgrading from v7.1 and previous versions.
- Other components
  - During the upgrade process, the operation of some components is not supported, such as:

>     - BR, Lightning ( Physical Import Mode ) : The operation of this type of component cannot be processed during the upgrade process, because the DDL in the paused state may be copied to TiDB during these operations. However, the DDL in this state cannot be automatically resumed, which may lead to subsequent DDL block.

Reviewer 2: ```suggestion
    - Backup/Restore, Lightning ( Physical Import Mode ) : The operation of this type of component cannot be processed during the upgrade process, because the DDL in the paused state may be copied to TiDB during these operations. However, the DDL in this state cannot be automatically resumed, which may lead to subsequent DDL block.
```

# Proposal: Pause User DDL when Upgrading TiDB

* Author: [zimulala](https://github.com/zimulala)
* Tracking issue: https://github.com/pingcap/tidb/issues/39751

> ## Abstract

Reviewer 2: Should add a blank line after each title?

> This document is designed to stop the execution of user DDL during the upgrade process. During the upgrade process, the user DDL operation is stopped first, then the system DDL operation during the upgrade process is completed, and finally, the original DDL operation is resumed (this function only considers the upgrade operation from v7.1 and later versions, that is, it does not support the upgrade from v7.1 upgrades from previous versions).

Reviewer 2: ```suggestion
This document describes a feature that allows users to pause the execution of their Data Definition Language (DDL) statements during the upgrade process of TiDB. During the upgrade process, the system stops the execution of user DDL statements, completes the system DDL operation, and finally resumes the execution of the original user DDL statements. This feature is available for upgrades from v7.1 and later versions of TiDB, and it does not support upgrades from versions prior to v7.1. 
```
